### ISSUE TITLE
question about library type stubs

### ISSUE THREAD

**Issue by cshesse** [microsoft/pyright#377]
The docs recommend putting all public type information for a library into a stub file, but it's unclear to me what the best practices here are.

I have ended up copying and pasting my type information from the individual files within the library into `__init__.pyi`, but I'm concerned that it will get out of sync with the "canonical" information stored on the original definitions.  Here are some possible ideas:

1) Remove type information from the original definitions so it only exists in one place.
2) Export type information automatically instead of manually (like wrapping each exported type with `# START EXPORT` and `# END EXPORT` and then running a script).
3) Check stubs against actual code (if the actual code is not annotated, just do the best you can, otherwise make sure everything matches exactly).

1 is slightly weird because you still have to remember that while private functions have their types directly on them, public ones need to have some other file updated.
2 is simple, if a little hacky
3 sounds like it could be annoying to do

I've already found 3 errors related to this in library code:

* https://github.com/pytorch/pytorch/pull/30441/files
* https://github.com/python/typeshed/pull/3479/files
* https://github.com/python/typeshed/pull/3484/files

For 2 of those at least, it seems like they could have been detected if there was a way to check stubs against actual code.

Is there a recommended way to do this?

**Comment by cshesse** [comment]
It looks like if I restructure my `.pyi` files a bit (`__init__.pyi` imports from some other `.pyi` files that are aligned with each `.py` file) I am able to use https://github.com/ambv/retype to check the types.

**Comment by erictraut** [comment]
There's no great answer to your question currently. I agree with you that there are problems with all of these approaches. 

I wouldn't recommend option 1 — for the reason you point out. 

I think options 2 and 3 are reasonable approaches, assuming that the tooling support them. It would be great to get agreement across the python community about the way to specify which symbols should be exported, for example. All tools could then honor that standard.

I think option 2 is the most elegant. It would allow library authors to write and maintain only one copy of their code and then generate high-quality type stubs automatically as part of the build process. This eliminates a bunch of maintenance and versioning issues associated with other approaches.

**Comment by cshesse** [comment]
After trying retype out, I immediately ran into two issues:
* https://github.com/ambv/retype/issues/23
* https://github.com/ambv/retype/issues/24

Thanks for your feedback!  I think I will give option 2 a try.

**Comment by cshesse** [comment]
I'm trying a fancier version of option 2: https://gist.github.com/cshesse/fd3bff91a84da0bfee63302f0c0ba4af

The idea is that `__init__.py` defines the public interface of each package (the exported symbols from each module), and must contain relative imports of the desired symbols.  `__init__.py` is copied to `__init__.pyi` and then each module in the package has its exported symbols copied to `<module>.pyi`.

This mostly seems to work (using python 3.8).  The nice thing is that this takes advantage of existing python conventions around `__init__.py` and while it doesn't work for arbitrary packages, a user should be able to restructure `__init__.py` to match the desired format.

The current hacky script is only sufficient for the one library that I am using it on, but a better version could be useful to avoid duplicating type information in stubs.

**Comment by cshesse** [comment]
Actually, after doing this it's not clear this works as I wanted.  It looks like pyright will see the pyi files, which are only the public interface, even when running pyright on the library itself, rather than on code that uses the library.

I guess this means I should only add pyi files when packaging the library for release?

**Comment by erictraut** [comment]
Yeah, do you think it's reasonable to do this as part of packaging the library for release, assuming it can be automated as a build step?

Out of curiosity, what library is this?

**Comment by cshesse** [comment]
It's a clone of tensorflow.gfile called `blobfile`, which is useful if you want to stream data from GCS/Azure Storage when using a non-tensorflow ML framework.

Mainly I am experimenting with type checking on this library to see if I want to use it for other projects.  I've experimented with it on other projects: https://github.com/openai/procgen/blob/master/procgen/libenv.py#L35 but not taken it very seriously.

So here's a summary of the approach for future reference, since I think I'm done messing with this for now:

1) The public API of your package should be all the stuff relative-imported in `__init__.py` (these are the exported symbols)
2) At build time, copy type information for only the exported symbols into .pyi files corresponding to each module and package them into the wheel.
   * This can be done with a setuptools command https://github.com/cshesse/blobfile/blob/master/setup.py#L12
   * Running `pip wheel . --no-deps` will cause this command to be run.
   * This requires a script like `export-stubs.py`, there's a [pyproject.toml](https://github.com/cshesse/blobfile/blob/master/pyproject.toml) to setup dependencies for that
   * You must also include `*.pyi` in `package_data` and probably add `py.typed` and set `zip_safe=False` for good measure.

Current limitations:

1) You must import the type symbols into your normal code, so you may want to import `Protocol` but to support python 3.6 you can't have that.  According to PEP 563 you can do `from __future__ import annotations` but that only works in python 3.7+.
2) The `export-stubs.py` script is hacky
    * doesn't handle a bunch of stuff, like TYPE_CHECKING
    * The script requires python 3.8
    * It should probably filter out any special logic from `__init__.py` and only look at the import statements

If this turns out to be a good idea then it may be worth having a not hacky version of `export-stubs.py` or a `pyright --exportstub <package>` command that does this.

Thanks again for all your help!
